### PR TITLE
Added Support for User Family Members

### LINKED ISSUE #1584: Add Support for User Family Members
**Is your feature request related to a problem? Please describe.**

1. To be even more useful to Administrators, there needs to be the ability to group Users as members of the same family.
2. When implemented in the API, the Admin Portal will be able to visualize and edit the family group members of each user in the user profile page.

**Describe the solution you'd like**
1. We should have the ability to add, and remove people from families. 
1. Families can only be valid if there are more than one member. Therefore when:
    1. there is only one member in a group, the family group entry should be empty
    2. a new family is created, the family group instantaneously has two family members
1. Other functionality must not be affected
1. You must write tests to validate your work

**Describe alternatives you've considered**
- N/A

**Approach to be followed (optional)**
- N/A

**Additional context**
- Work with @xoldyckk on evaluating your solution here and in the subsequent PR
- This may be a prerequisite for our menu of project ideas for GSoC 2024

**Potential internship candidates**
- Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by AdityaRaimec22**
**What kind of change does this PR introduce?**
Add Support for User Family Members

**Issue Number:**

Fixes #1584 

**Did you add tests for your changes?**
Yes

**Snapshots/Videos:**

**If relevant, did you update the documentation?**
Not relevant

**Summary**

**Does this PR introduce a breaking change?**
No

**Other information**
Some tests are failing and I am currently addressing them, and I have raised a pull request because the assigned issue has been pending for a significant period.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
Please fix the conflicting files

File: tests/resolvers/Mutation/createFamilyGroup.spec.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import "dotenv/config";
+import type mongoose from "mongoose";
+import { Types } from "mongoose";
+import type { MutationCreateFamilyGroupArgs } from "../../../src/types/generatedGraphQLTypes";
```
Comment (palisadoes) [inline tests/resolvers/Mutation/createFamilyGroup.spec.ts:4]: You have interchangeably used the term `familyGroup` and `Family` in the PR definitions. To make it unambiguous for future developers use `userFamily`  to replace `familyGroup` and `Family`.

**Review by palisadoes — CHANGES_REQUESTED** [review]
You have interchangeably used the term `familyGroup` and `Family` in the PR definitions. To make it unambiguous for future developers use `userFamily`  to replace `familyGroup` and `Family`.

File: tests/resolvers/Mutation/createFamilyGroup.spec.ts (hunk 1 above)
Comment (AdityaRaimec22) [reply tests/resolvers/Mutation/createFamilyGroup.spec.ts]: done

**Review by palisadoes — CHANGES_REQUESTED** [review]
The Mutation Test file names should more closely reflect the UserFamily naming convention

image

**Comment by AdityaRaimec22** [comment]
> The Mutation Test file names should more closely reflect the UserFamily naming convention
> 
> image

committed changes @palisadoes

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by AdityaRaimec22** [comment]
working on the failing tests will fix them ASAP.

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by palisadoes** [comment]
Please fix the conflicting file

**Comment by AdityaRaimec22** [comment]
@palisadoes The PR is up for review now all the tests has been passed from my side.

File: src/models/User.ts [hunk 2]
```diff
@@ -126,6 +130,12 @@ const userSchema = new Schema(
         ref: "Organization",
       },
     ],
+    adminForUserFamily: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Organization",
+      },
+    ],
```
Comment (EshaanAgg) [inline src/models/User.ts:138]: The field name and the `ref` don't match. Shouldn't it be `UserFamily`?
Also, there is no need to store the same data twice, both on the User Schema and the User Family schema. We should also store it once, from where the queries and mutations would access it more frequently.

File: src/models/User.ts [hunk 3]
```diff
@@ -140,6 +150,12 @@ const userSchema = new Schema(
         ref: "Organization",
       },
     ],
+    createdUserFamily: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Organization",
+      },
+    ],
```
Comment (EshaanAgg) [inline src/models/User.ts:158]: Same comment as earlier. Remove repetition.

File: src/models/User.ts [hunk 4]
```diff
@@ -200,6 +216,12 @@ const userSchema = new Schema(
         ref: "Organization",
       },
     ],
+    joinedUserFamily: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Organization",
+      },
+    ],
```
Comment (EshaanAgg) [inline src/models/User.ts:224]: Same as above. Remove repetition.

File: src/models/userFamily.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      type: Schema.Types.ObjectId,
+      ref: "User",
+      required: true,
+    },
+  ],
+  creator: {
+    type: Schema.Types.ObjectId,
+    ref: "User",
+    required: true,
+  },
```
Comment (EshaanAgg) [inline src/models/userFamily.ts:48]: This data is repeated and adds complexity to the resolvers to maintain integrity. Either store them here or on the user object. According to me these all are better stored here, but Im open to counter arguments.

**Review by EshaanAgg — CHANGES_REQUESTED** [review]
Need some changes in the proposed schema, and subsequently, the implemented resolvers. Otherwise, the overall implementation looks good.

File: src/models/userFamily.ts (hunk 5 above)
Comment (AdityaRaimec22) [reply src/models/userFamily.ts:48]: Committed all the asked changes the PR is up for review now. Thanks for suggestions.

**Comment by SiddheshKukade** [comment]
@AdityaRaimec22 where is the UI Equivalent fort this feature can you please link that I wanted to check that.

File: src/resolvers/Mutation/adminRemoveMemberFromUserFamily.ts
[the remark this answers is not in the document]
Comment (AdityaRaimec22) [reply src/resolvers/Mutation/adminRemoveMemberFromUserFamily.ts:1]: done

File: src/resolvers/Mutation/adminAddMemberToUserFamily.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    //check whether user is admin of the family
+    await adminCheck(currentUser?._id, userFamily);
+
+    const isUserMemberOfUserFamily = userFamily.users.some((user) => {
+      user.equals(args.userId);
+    });
+
+    // Checks whether user with _id === args.userId is already a member of Family.
+    if (isUserMemberOfUserFamily === true) {
```
[the remark this answers is not in the document]
Comment (AdityaRaimec22) [reply src/resolvers/Mutation/adminAddMemberToUserFamily.ts]: done

File: src/resolvers/Mutation/adminAddMemberToUserFamily.ts
[the remark this answers is not in the document]
Comment (AdityaRaimec22) [reply src/resolvers/Mutation/adminAddMemberToUserFamily.ts:1]: done

**Review by EshaanAgg — APPROVED** [review]

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by AdityaRaimec22** [comment]
@palisadoes @EshaanAgg All the checks are passed from my side. Can you approve the PR now. Thanks

image

**Comment by AdityaRaimec22** [comment]
> @AdityaRaimec22 where is the UI Equivalent fort this feature can you please link that I wanted to check that.

@SiddheshKukade here is the UI equivalent of this feature.


https://github.com/PalisadoesFoundation/talawa-api/assets/118007662/8d2ad07b-a77a-4a60-bbfd-ba91ab94ee50

**Review by EshaanAgg — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]

**Comment by palisadoes** [comment]
@AdityaRaimec22 

You bypassed the lint rules some how and now everyone else is having linting issues. Please resubmit your changes with the correct linting applied.

**Comment by AdityaRaimec22** [comment]
@palisadoesI haven't overridden or bypassed any of the changes; they were all pushed in the usual manner, and the checks also passed in this pull request. Could you please provide a link to a pull request where the lint checks failed, or share some resource indicating that checks failed for this reason? I'd like to take a look at why those changes failed.

**Comment by palisadoes** [comment]
@AdityaRaimec22 

1. That was the impression given to me by @LakshyaSatpal by this comment right after i merged. When it was reverted the issue went away.
2. However, that may have been caused by fixing merge conflicts which may have triggered the problem, based on the fact that he mentioned the conflicts going away, which were not mentioned initially.  It could have been a question of timing as it was noted very soon after the merge.
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1774#issuecomment-1925459699
If you feel confident that the linting was OK, let's resubmit and let's try to see whether others have the same issue.

I'll hang around until you do.

**Comment by palisadoes** [comment]
If I was hasty, I apologize. The timing was so close that I might have jumped to an incorrect conclusion

**Comment by AdityaRaimec22** [comment]
> If I was hasty, I apologize. The timing was so close that I might have jumped to an incorrect conclusion

@palisadoes so should I open a new PR.

**Comment by palisadoes** [comment]
Yes

**Comment by LakshyaSatpal** [comment]
@AdityaRaimec22 In most files, the use of `any` in your changes, was causing lint errors.

**Comment by AdityaRaimec22** [comment]
> @AdityaRaimec22 In most cases, the use of `any` in your changes, was causing lint errors. I confirmed this by looking at the changed files in this PR.

@LakshyaSatpal The code I've added isn't failing in the checks of this PR or when tested locally. Additionally, the code I've implemented draws inspiration from existing code in the codebase; I haven't introduced any new methods to incorporate this feature. It is due to other reasons which we are working on.

**Pull request merged**